File: src/go/problems/make_equal_test.go
// https://leetcode.com/problems/redistribute-characters-to-make-all-strings-equal

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type MakeEqual struct{}

func TestMakeEqual(t *testing.T) {
	gen := core.TestSuite[MakeEqual]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([]string{"abc", "aabc", "bc"}).Result(true)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]string{"ab", "a"}).Result(false)
	}).Add(func(tc *core.TestCase) {
		tc.Param([]string{"a", "b"}).Result(false)
	}).Run(t)
}

func (MakeEqual) Solution(words []string) bool {
	set := make([]int, 26)

	for _, word := range words {
		for _, ch := range word {
			set[ch-'a']++
		}
	}

	length := len(words)
	for _, count := range set {
		if count%length != 0 {
			return false
		}
	}

	return true
}

File: src/python/problems/make_equal_test.py
# https://leetcode.com/problems/redistribute-characters-to-make-all-strings-equal

from core.problem_base import *

class MakeEqual(ProblemBase):
    def Solution(self, words: list[str]) -> bool:
        map = [0] * 26

        for word in words:
            for ch in word:
                map[ord(ch) - ord('a')] += 1

        length = len(words)
        for count in map:
            if count % length != 0:
                return False

        return True

if __name__ == '__main__':
    TestGen(MakeEqual) \
        .Add(lambda tc: tc.Param(["abc","aabc","bc"]).Result(True)) \
        .Add(lambda tc: tc.Param(["ab","a"]).Result(False)) \
        .Add(lambda tc: tc.Param(["a","b"]).Result(False)) \
        .Run()
